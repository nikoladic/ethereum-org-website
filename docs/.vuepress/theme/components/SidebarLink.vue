<script>
import {
  isActive,
  hashRE,
  groupHeaders,
  resolveHeaderTitle
} from '../utils/util'

export default {
  functional: true,

  props: ['item'],

  render(h, { parent: { $page, $site, $route }, props: { item } }) {
    // use custom active class matching logic
    // due to edge case of paths ending with / + hash
    const selfActive = isActive($route, item.path)
    // for sidebar: auto pages, a hash link should be active if one of its child
    // matches
    const active =
      item.type === 'auto'
        ? selfActive ||
          item.children.some(c =>
            isActive($route, item.basePath + '#' + c.slug)
          )
        : selfActive
    const link = renderLink(h, item.path, item.title || item.path, active)
    const configDepth =
      $page.frontmatter.sidebarDepth != null
        ? $page.frontmatter.sidebarDepth
        : $site.themeConfig.sidebarDepth
    const maxDepth = configDepth == null ? 1 : configDepth
    const displayAllHeaders = !!$site.themeConfig.displayAllHeaders
    if (item.type === 'auto') {
      return h('div', {}, [
        link,
        renderChildren(h, item.children, item.basePath, $route, maxDepth)
      ])
    } else if (
      (active || displayAllHeaders) &&
      item.headers &&
      !hashRE.test(item.path)
    ) {
      const children = groupHeaders(item.headers)
      return h('div', {}, [
        link,
        renderChildren(h, children, item.path, $route, maxDepth)
      ])
    } else {
      return link
    }
  }
}

function renderLink(h, to, text, active, isSubHeader = false) {
  return h(
    'a',
    {
      attrs: {
        href: to
      },
      class: {
        'tc-primary500 active': active,
        'bc-primary500': active && !isSubHeader,
        'tc-text100': !active,
        'sidebar-link br-25rem w-100 l8 tc-h-primary500 mt-0 mb-05 pr-025': true
      }
    },
    resolveHeaderTitle(text)
  )
}

function renderChildren(h, children, path, route, maxDepth, depth = 1) {
  if (!children || depth > maxDepth) return null
  return h(
    'ul',
    { class: 'pl-1' },
    children.map(c => {
      const active = isActive(route, path + '#' + c.slug)
      return h('li', { class: 'pl-1' }, [
        renderLink(h, path + '#' + c.slug, c.title, active, true),
        renderChildren(h, c.children, path, route, maxDepth, depth + 1)
      ])
    })
  )
}
</script>
