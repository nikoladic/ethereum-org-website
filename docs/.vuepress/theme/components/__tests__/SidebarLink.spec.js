import { shallowMount } from '@vue/test-utils'
import SidebarLink from "@/theme/components/SidebarLink"

jest.mock("../../../theme/utils/util");

describe('SidebarLink', () => {

  let config = {
    propsData: {
      item: {
        basePath: "/",
        path: "/",
        children: [],
        title: "No Children",
        headers: ["header1", "header2", "header3"]
      }
    },
    context: {
      parent: {
        $route: {}
      }
    },
    mocks: {
      $page: {
        frontmatter: {
          sidebarDepth: 2
        }
      },
      $site: {
        themeConfig: {
          sidebarDepth: 2,
          displayAllHeaders: true
        }
      }
    },
    stubs: ['router-link']
  }

  test('without children and headers', () => {
    const configCopy = Object.assign({}, config);
    configCopy.propsData.item = {
      basePath: "/",
      path: "/",
      children: [],
      title: "No Children and Headers",
    }
    const wrapper = shallowMount(SidebarLink, configCopy)
    expect(wrapper.element).toMatchSnapshot()
  })

  test('with headers but no children', () => {
    const configCopy = Object.assign({}, config);
    configCopy.propsData.item = {
      basePath: "/",
      path: "/",
      children: [],
      title: "No Children",
      headers: [{
          level: 1,
          title: "Header 1",
          slug: "header1"
        }, {
          level: 2,
          title: "Header 2",
          slug: "header2"
        }, {
          level: 2,
          title: "Header 3",
          slug: "header3"
        }]
    }
    const wrapper = shallowMount(SidebarLink, configCopy)
    expect(wrapper.element).toMatchSnapshot()
  })

  test('with children in auto mode', () => {
    const configCopy = Object.assign({}, config);
    configCopy.propsData.item = {
      type: "auto",
      basePath: "/",
      path: "/",
      children: [{
        level: 2,
        title: "Child 1",
        slug: "child2"
        }, {
        level: 2,
        title: "Child 2",
        slug: "child2"
      }],
      title: "With Children",
    }
    const wrapper = shallowMount(SidebarLink, configCopy)
    expect(wrapper.element).toMatchSnapshot()
  })
})

