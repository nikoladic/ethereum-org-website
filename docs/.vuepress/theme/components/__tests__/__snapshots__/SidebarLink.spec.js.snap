// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarLink with children in auto mode 1`] = `<!---->`;

exports[`SidebarLink with headers but no children 1`] = `<!---->`;

exports[`SidebarLink without children and headers 1`] = `
<router-link-stub
  class="active sidebar-link"
>
  No Children and Headers
</router-link-stub>
`;
